psa-crypto/src/operations/key_derivation.rs

```diff
@@ -2,3 +2,95 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! # Key Derivation operations
+
+use super::key_management::complete_new_key_operation;
+use crate::initialized;
+use crate::types::key::Attributes;
+use crate::types::key::Id;
+use crate::types::key_derivation_operation::KeyDerivationOperation;
+use crate::types::status::{Result, Status};
+use core::convert::{TryFrom, TryInto};
+
+/// This function calculates output bytes from a key derivation algorithm and uses those bytes to generate a key deterministically.
+/// The key's location, usage policy, type and size are taken from attributes.
+/// # Example
+///
+/// ```
+/// use psa_crypto::operations::{key_derivation, key_management};
+/// use psa_crypto::types::key::{Attributes, Type, Lifetime, Policy, UsageFlags};
+/// use psa_crypto::types::algorithm::{Hash, KeyDerivation};
+/// use psa_crypto::types::key_derivation_operation::{KeyDerivationOperation, KeyDerivationWithInputs, Input};
+///
+/// # const KEY_DATA: [u8; 23] = [0; 23];
+/// let mut attributes = Attributes {
+///     key_type: Type::Derive,
+///     bits: 0,
+///     lifetime: Lifetime::Volatile,
+///     policy: Policy {
+///         usage_flags: UsageFlags {
+///             derive: true,
+///             ..Default::default()
+///         },
+///         permitted_algorithms: KeyDerivation::Hkdf {
+///            hash_alg: Hash::Sha256,
+///         }.into()
+///     }
+/// };
+///
+/// let mut derived_key_attributes = Attributes {
+/// key_type: Type::RawData,
+///     bits: 8,
+///     lifetime: Lifetime::Volatile,
+///     policy: Policy {
+///         usage_flags: UsageFlags {
+///             derive: true,
+///             ..Default::default()
+///         },
+///         permitted_algorithms: KeyDerivation::Hkdf {
+///            hash_alg: Hash::Sha256,
+///         }.into()
+///     }
+/// };
+///
+/// psa_crypto::init().unwrap();
+/// let my_key = key_management::import(attributes, None, &KEY_DATA).unwrap();
+/// let info = vec![20; 0x3f];
+/// let mut operation = KeyDerivationOperation {
+///     alg: KeyDerivationWithInputs::Hkdf {
+///         hash_alg: Hash::Sha256,
+///         salt: None,
+///         secret: Input::Key(my_key),
+///         info: Input::Bytes(&info),
+/// },
+///     capacity: None,
+/// };
+/// let _new_key = key_derivation::key_derivation_operation_output_key(operation, derived_key_attributes, None).unwrap();
+/// ```
+pub fn key_derivation_operation_output_key(
+    operation: KeyDerivationOperation,
+    attributes: Attributes,
+    id: Option<u32>,
+) -> Result<Id> {
+    initialized()?;
+
+    let mut key_attributes = psa_crypto_sys::psa_key_attributes_t::try_from(attributes)?;
+    let id = if let Some(id) = id {
+        unsafe { psa_crypto_sys::psa_set_key_id(&mut key_attributes, id) };
+        id
+    } else {
+        0
+    };
+    let mut handle_for_new_key = 0;
+
+    let mut op: psa_crypto_sys::psa_key_derivation_operation_t = operation.clone().try_into()?;
```

Comment:
The reason for not implementing `Copy` on this was to stop the user ending up with an `operation` that they could use after it was aborted. The current way it is setup, once the user calls the key derivation operation, they cannot use that `operation` again.

When you call abort of the C struct, it resets it to default values, which are not 'valid'. I didn't want to make the types on `operation` `Option<T>` as I don't think it fits, so there is no way to reset the Rust `operation` struct, you have to start again, specifying the inputs you want at each stage for whatever algorithm you're using.

Another way of doing it would be to allow the user to keep their `operation` struct and re-use it. Its just with the `abort` in Mbed Crypto, I was trying to replicate it. Now that I think about it, I can't think of a reason why it shouldn't be done like that.

This is more of a Rust thing but I find that `clone` a bit annoying. I played around with making the `TryFrom` method take a reference, but then it got upset when I had the `abort` there as there was still a living borrow. I get why its complaining, but I know the borrow isn't used (e.g. stored) for use after `TryFrom` returns!

Reply:
> to stop the user ending up with an operation that they could use after it was aborted.

They can still clone it, right? An alternative is to implement your own internal clone that's only visible to this crate.

Reply:
> Or a TryFrom<&KeyDerivationOperation> if that's possible, I think that's a thing

That's what I was referring to in this comment. You can have a `TryFrom` with a borrow, but that doesn't solve this problem ☹️.

> This is more of a Rust thing but I find that clone a bit annoying. I played around with making the TryFrom method take a reference, but then it got upset when I had the abort there as there was still a living borrow. I get why its complaining, but I know the borrow isn't used (e.g. stored) for use after TryFrom returns!

Reply:
What will  happen for the slices when cloning? Will that make deep copies of them?

Reply:
Hmmm good question. Slice [doesn't seem to implement](https://doc.rust-lang.org/std/primitive.slice.html#trait-implementations) `Clone`. It could copy the reference but then you could end up with a dangling pointer, which Rust doesn't allow 🤔.